events: Add new `user_settings` event type when updating user settings.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
There are four commits -
- Send `user_settings` event for both display and notification settings.
- Add `user_settings_object` to `client_capabilities`. The old events will be sent only is `user_settings_object` is false.
- Add `user_settings` object, containing all display and notification settings, to `register` response.
- Update frontned code to use `user_settings` event.

[Discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/events.20for.20notification.20and.20display.20settings)
 <!-- How have you tested? -->

 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on zerver/lib/events.py:
```diff
@@ -511,6 +511,20 @@ def fetch_initial_state_data(
+        state["user_settings"]["timezone"] = settings_user.timezone
+        state["user_settings"][
+            "available_notification_sounds"
+        ] = get_available_notification_sounds()
```
I am not sure if we should include `emojiset_choices` and `available_notification_sounds` as they are not settings actually and are just the possible choices for these settings.

timabbott replied:
Good question; maybe one for `#api design`.  It kinda makes sense to send these whenever we're sending the `user_settings`, since they are the values that let one render UI for changing those values and are small, but I can see an argument for putting this sort of thing in a separate object.

sahil839 commented:
There will be some changes, mainly in api docs and some merge conflicts in changelog, when #19394 is merged.

sahil839 commented:
Have also asked the question of two copies of settings values in the register response in chat.zulip.org.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -243,6 +243,10 @@ paths:
                                 have changed.
+
+                                **Changes**: Deprecated in Zulip 5.0 (feature level 87), replaced by
+                                the `user_settings` event type
```
Sentences should end with a `.`.

sahil839 replied:
Done this.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -307,6 +315,51 @@ paths:
+                                have changed.
+
+                                **Changes**: New in Zulip 5.0 (feature level 87), replacing the
+                                previous `update_display_settings` event type, which is still
```
This should mention notification settings too, right?

sahil839 replied:
Yeah missed that, updated now.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -307,6 +315,51 @@ paths:
+                                  "type": "user_settings",
+                                  "setting_name": "high_contrast_mode",
+                                  "setting": false,
+                                  "id": 0,
```
Do we want to adjust the format for these dictionaries as well?  E.g. I think `/setting/value/` would definitely be an improvement; might be worth doing `s/setting_name/property/` to match the `do_set_realm_property` format.

sahil839 replied:
Yes, this would be better. Updated.

timabbott commented on zerver/lib/events.py:
```diff
@@ -1061,6 +1061,13 @@ def apply_event(
         assert event["notification_name"] in UserProfile.notification_setting_types
         state[event["notification_name"]] = event["setting"]
+    elif event["type"] == "user_settings":
+        if event["setting_name"] != "timezone":
```
Let's add a comment documenting why this exception exists.

sahil839 replied:
Added the comment.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -7762,6 +7767,13 @@ paths:
               future server release.
+
+            * `user_settings_object`: Boolean for whether the `update_display_settings`
+              and `update_global_notifications` events are sent to clients.
```
This explanation is too terse; we should explain clearly that those events are intended to be replaced by the new `user_settings` event type. Maybe:

"Boolean for whether the client supports the modern `user_settings` event type. If False,
the server will additionally send the legacy `update_display_settings` and `update_global_notifications` 
event types for backwards-compatibility with clients that predate this API migration.

..."

sahil839 replied:
Updated the text with this. Thanks for the suggestion.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -217,6 +221,11 @@ def accepts_event(self, event: Mapping[str, Any]) -> bool:
+            # 'update_display_settings' and 'update_global_notifications'
+            # events are sent only if user_settings_object is False,
+            # otherwise only 'user_settings' are sent.
+            return not self.user_settings_object
```
I think where we landed with how this should work is that we can put this at the top as:
```
if not self.user_settings_object and event["type"] in ["update_display_settings", "update_global_notifications"]:
    return False
```

sahil839 replied:
@timabbott  As per the chat in chat.zulip.org, 
> We add a user_settings_object entry to client_capabilities, which if sent, results in the update_global_notifications 
> and update_display_settings objects being ignored. (We need this because of how some clients request "all event > types", and we'd like the "all" to be able to not include the deprecated legacy types at all on a modern client).

which means that this should return false (i.e. not accept the event) if `user_settings_object` is true and this makes sense with the variable name `user_settings_object` as well, I think, I will wait for the reply before changing this.

timabbott replied:
Yes, sorry I meant:
```
if self.user_settings_object and event["type"] in ["update_display_settings", "update_global_notifications"]:
    return False
```

which is what you described; the `not` there was a typo.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8474,6 +8474,233 @@ paths:
+                        description: |
+                          Present if `user_settings` is present in `fetch_event_types`.
+
+                          A dictionary which contains the settings of a user.
```
"A dictionary containing the user's personal settings."

sahil839 replied:
Done.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8474,6 +8474,233 @@ paths:
+
+                          A dictionary which contains the settings of a user.
+
+                          **Changes**: New in Zulip 5.0 (feature level 89).
```
This should have a 

"Previously, these settings appeared in the top-level object, where they are available
for clients without the `user_settings_object` client capability for backwards-compatibility.

sahil839 replied:
Done.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8474,6 +8474,233 @@ paths:
+                          twenty_four_hour_time:
+                            type: boolean
+                            description: |
+                              Whether time should be [displayed in 24-hour notation](/help/change-the-time-format).
```
I'm not reviewing the individual settings text, under the assumption that it's just moved.

sahil839 replied:
Yes, have copied it. I was not able to figure out how to avoid copying, especially with the `Changes` entries here in the `/settings` endpoint documentation.

timabbott replied:
Yeah, the `Changes` system often ends up forcing copying, which is annoying, but not much to be done about it.

timabbott commented on static/js/server_events_dispatch.js:
```diff
@@ -539,7 +539,17 @@ export function dispatch_normal_event(event) {
 
-        case "update_display_settings": {
+        case "user_settings": {
+            if (settings_config.all_notification_settings.includes(event.setting_name)) {
```
How did you verify that `all_notification_settings` lists all notification setting event types?

sahil839 replied:
I checked that the settings in `all_notification_settings` and in `notification_setting_types` dict, except the `presence_enabled` and `email_notifications_batching_period_seconds` which is only in the `notification_setting_types` dict and we don't have any live update code for `presence_enabled` and `email_notifications_batching_period_seconds` is implemented in backend only.

timabbott replied:
OK, makes sense. I expect it's a bug that `presence_enabled` does not have live-update code, hrm.

timabbott commented:
Thanks for working on this @sahil839!  I think this PR is pretty close to being mergeable; we mostly just need to finalize these last bits of API semantics and make sure the API documentation is clear about how it works.

sahil839 commented:
Yes, would be better if we resume this after doing that migration.

sahil839 commented:
@timabbott updated the PR with suggested changes except [this](https://github.com/zulip/zulip/pull/19404#discussion_r678667277) for which I have commented in the thread itself.
Also, I have almost done the work of controling the setting fields in top-level object using `client_capabilities` and the frontend changes for using the `user_settings` object, but need to test it carefully, Will complete this by tommorrow.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5065,6 +5065,14 @@ def do_change_notification_settings(
 
     send_event(user_profile.realm, event, [user_profile.id])
 
+    event = {
```
I renamed these local variables to `legacy_event`, just for readability reasons, and also to avoid having the `event` name have two types in these functions.

timabbott commented:
This PR looks great!  I pushed back after the small variable name change noted above and also keeping all the changes at feature level 87, since I think it makes sense for this to be an atomic update for the purposes of clients.

I'll hold off on merging until you add that last commit which can also be at feature level 87; please comment if you need to make any changes to the existing commits, since that way I can avoid rereading the ones I've already reviewed and am happy with.

timabbott commented:
I just merged #18074, so this will need a feature level bump.

timabbott commented on static/js/composebox_typeahead.js:
```diff
@@ -1092,7 +1092,7 @@ export function initialize() {
 
     $("#enter_sends").on("click", () => {
-        page_params.enter_sends = $("#enter_sends").is(":checked");
+        page_params.user_settings.enter_sends = $("#enter_sends").is(":checked");
```
Reading this, I wonder if we want to create a `user_settings` module to hold these settings rather than accessing them through `page_params`; we can just make `ui_init.js` initialize the `user_settings` module with the `page_prams.user_settings` object and then everything can import and access that module.  It'd certainly make code like this read more cleanly, and I think would help a lot in reducing the amount of random code that needs to interact with `page_params` at all.

Making a `user_settings` module, which would be just a dictionary with all the settings in it, like present `page_params` module, might be a pretty clean preparatory refactor we do on master, if you think it's worth doing.

sahil839 replied:
Yeah I agree. We could do it as a preparatory PR, will be easy to review this PR I think. 
We should add a typescript module, right?

sahil839 commented:
This is not ready for review yet as I have to modify the code for the recently added settings like `enable_drafts_synchronization`.

sahil839 commented:
@timabbott Updated the PR.
- Added a prep commit for removing duplicate tests for `enable_drafts_synchronization`.
- Updated the feature levels in previous commits and also added `enable_drafts_synchronization` in the api docs of `user_settings` object.
- There is one change in previous commits - moved the change to make `user_settings_object` True to the last commit of adding the `user_settings` module. Rest, the old commits are same.
- I added the `user_settings` module in this PR only, as I was facing some issues with lint for typescript and it would have resulted in some unnecessary changes which would have been removed anyway in this PR.

sahil839 commented on frontend_tests/zjsunit/zpage_params.js:
```diff
@@ -1,11 +1,17 @@
+        if (Object.prototype.hasOwnProperty.call(exports.user_settings, field)) {
+            delete exports.user_settings[field];
+        }
+    }
```
I added the `user_settings` code here as it works similarly to `page_params` in the node tests. Not sure this change is worth a separate file or we should change this file name.

timabbott replied:
This looks great!

timabbott commented:
Merged, after adding a commit adding complete types to the `user_settings` object; @priyank-p FYI.

I also did a bit more work on the documentation in a final commit.  Huge thanks for executing this transition @sahil839!
